File: scripts/video-to-ascii.mjs
#!/usr/bin/env node
/**
 * Convert video to ASCII frames JSON for web playback
 * Inspired by Ghostty's video-to-terminal approach
 *
 * Usage: node scripts/video-to-ascii.mjs <input.mp4> [output.json] [cols] [fps]
 */

import { execSync, spawn } from 'child_process'
import { createCanvas, loadImage } from 'canvas'
import { writeFileSync, mkdirSync, rmSync, readdirSync } from 'fs'
import { join } from 'path'
import { tmpdir } from 'os'

const ASCII_CHARS = ' .·:;=+x%#@'

const inputFile = process.argv[2]
const outputFile = process.argv[3] || 'public/ascii-frames.json'
const cols = parseInt(process.argv[4]) || 80
const fps = parseInt(process.argv[5]) || 12

if (!inputFile) {
  console.log('Usage: node scripts/video-to-ascii.mjs <input.mp4> [output.json] [cols] [fps]')
  process.exit(1)
}

// Create temp directory
const workDir = join(tmpdir(), `ascii-${Date.now()}`)
mkdirSync(workDir, { recursive: true })

console.log(`Converting ${inputFile} to ASCII...`)
console.log(`Settings: ${cols} columns, ${fps} fps`)

try {
  // Extract frames with ffmpeg
  console.log('Extracting frames...')
  execSync(
    `ffmpeg -i "${inputFile}" -vf "scale=${cols}:-1:flags=lanczos" -r ${fps} "${workDir}/frame_%04d.png" -loglevel warning`,
    { stdio: 'inherit' }
  )

  // Get list of frames
  const frames = readdirSync(workDir)
    .filter(f => f.startsWith('frame_'))
    .sort()

  console.log(`Processing ${frames.length} frames...`)

  const asciiFrames = []
  const canvas = createCanvas(cols, 100) // Height will be adjusted per frame
  const ctx = canvas.getContext('2d')

  for (let i = 0; i < frames.length; i++) {
    const img = await loadImage(join(workDir, frames[i]))

    // Adjust canvas to match image
    const rows = img.height
    canvas.width = cols
    canvas.height = rows

    // Draw image
    ctx.drawImage(img, 0, 0, cols, rows)

    // Get pixel data
    const imageData = ctx.getImageData(0, 0, cols, rows)
    const pixels = imageData.data

    // Convert to ASCII
    let ascii = ''
    for (let y = 0; y < rows; y++) {
      for (let x = 0; x < cols; x++) {
        const idx = (y * cols + x) * 4
        const r = pixels[idx]
        const g = pixels[idx + 1]
        const b = pixels[idx + 2]

        // Luminance
        const lum = 0.2126 * r + 0.7152 * g + 0.0722 * b
        const charIdx = Math.floor((lum / 255) * (ASCII_CHARS.length - 1))
        ascii += ASCII_CHARS[charIdx]
      }
      ascii += '\n'
    }

    asciiFrames.push(ascii.trimEnd())
    process.stdout.write(`\rProcessed ${i + 1}/${frames.length}`)
  }

  console.log('\nWriting output...')

  const output = {
    fps,
    cols,
    frameCount: asciiFrames.length,
    frames: asciiFrames
  }

  writeFileSync(outputFile, JSON.stringify(output))
  console.log(`Done! Output: ${outputFile}`)
  console.log(`File size: ${(JSON.stringify(output).length / 1024).toFixed(1)} KB`)

} finally {
  // Cleanup
  rmSync(workDir, { recursive: true, force: true })
}

File: scripts/video-to-ascii.py
#!/usr/bin/env python3
"""
Convert video to ASCII frames JSON for web playback
Much faster than bash version - uses numpy for pixel processing
"""

import subprocess
import json
import sys
import tempfile
import os

# Rich ASCII character set from dark to light
ASCII_CHARS = ' .·:+*oø®œ#@'

def main():
    if len(sys.argv) < 2:
        print("Usage: python video-to-ascii.py <input.mp4> [output.json] [cols] [fps]")
        sys.exit(1)

    input_file = sys.argv[1]
    output_file = sys.argv[2] if len(sys.argv) > 2 else 'public/ascii-frames.json'
    cols = int(sys.argv[3]) if len(sys.argv) > 3 else 160
    fps = int(sys.argv[4]) if len(sys.argv) > 4 else 12

    print(f"Converting {input_file} to ASCII...")
    print(f"Settings: {cols} columns, {fps} fps")

    # Get video dimensions
    probe = subprocess.run([
        'ffprobe', '-v', 'error', '-select_streams', 'v:0',
        '-show_entries', 'stream=width,height',
        '-of', 'csv=p=0', input_file
    ], capture_output=True, text=True)

    orig_width, orig_height = map(int, probe.stdout.strip().split(','))
    rows = int(orig_height * cols / orig_width / 2)  # /2 for terminal aspect ratio
    if rows < 1:
        rows = 1

    print(f"Frame dimensions: {cols}x{rows}")

    # Extract frames as raw grayscale
    with tempfile.NamedTemporaryFile(suffix='.raw', delete=False) as tmp:
        raw_file = tmp.name

    try:
        subprocess.run([
            'ffmpeg', '-i', input_file,
            '-vf', f'scale={cols}:{rows},format=gray',
            '-r', str(fps),
            '-f', 'rawvideo',
            '-pix_fmt', 'gray',
            raw_file,
            '-loglevel', 'warning', '-y'
        ], check=True)

        # Read raw data
        with open(raw_file, 'rb') as f:
            raw_data = f.read()

        frame_size = cols * rows
        frame_count = len(raw_data) // frame_size
        print(f"Processing {frame_count} frames...")

        frames = []
        num_chars = len(ASCII_CHARS)

        for f_idx in range(frame_count):
            offset = f_idx * frame_size
            frame_bytes = raw_data[offset:offset + frame_size]

            ascii_frame = []
            for row in range(rows):
                row_start = row * cols
                line = ''
                for col in range(cols):
                    gray = frame_bytes[row_start + col]
                    char_idx = gray * num_chars // 256
                    if char_idx >= num_chars:
                        char_idx = num_chars - 1
                    line += ASCII_CHARS[char_idx]
                ascii_frame.append(line)

            frames.append('\n'.join(ascii_frame))

            if f_idx % 10 == 0:
                print(f"\rProcessed {f_idx + 1}/{frame_count} frames...", end='', flush=True)

        print(f"\nWriting {output_file}...")

        output = {
            'fps': fps,
            'cols': cols,
            'rows': rows,
            'frameCount': len(frames),
            'frames': frames
        }

        with open(output_file, 'w') as f:
            json.dump(output, f)

        size_kb = os.path.getsize(output_file) / 1024
        print(f"Done! Output: {output_file} ({size_kb:.1f} KB)")

    finally:
        if os.path.exists(raw_file):
            os.unlink(raw_file)

if __name__ == '__main__':
    main()
